src/test-utils/browser-automation/browser-driver.js

```diff
@@ -18,137 +18,180 @@
   }
 
-  startServer(config = {}, maxRetryTimes = 30) {
-    const newConfig = Object.assign({}, DEFAULT_CONFIG, config);
-    return new Promise((resolve, reject) => {
-      log.log({
-        message: `Binding to port: ${newConfig.port}`,
-        color: COLOR.YELLOW
-      })();
-      const timeout = setTimeout(() => { // eslint-disable-line
-        resolve();
-      }, 2000);
-      this.child = this.execFile(
-        newConfig.process,
-        [...newConfig.parameters, '--port', `${newConfig.port}`],
-        newConfig.options,
-        error => {
-          if (error) {
-            clearTimeout(timeout); // eslint-disable-line
-            log.log({
-              message: `Failed to bind port: ${newConfig.port}`,
-              color: COLOR.YELLOW
-            })();
-            reject(error);
-          }
-        }
-      );
-    })
-      .then(() => {
-        this.port = newConfig.port;
-      })
-      .catch(error => {
-        if (maxRetryTimes > 0) {
-          newConfig.port++;
-          return this.startServer(newConfig, maxRetryTimes - 1);
-        } else { // eslint-disable-line
-          log.log({
-            message: 'Failed to start server, use \'killall node\' to stop existing services',
-            color: COLOR.RED
-          })();
-          throw error;
-        }
+  // Starts a web server with the provided configs
+  // Resolves to the bound url if successful
+  startServer(config = {}) {
+    config = mergeServerConfigs(config);
+
+    const getPort = config.port === 'auto' ?
+      this._getAvailablePort() : Promise.resolve(config.port);
+
+    return getPort.then(port => new Promise((resolve, reject) => {
+      const args = [...config.arguments];
+      if (port) {
+        args.push('--port', port);
+      }
+
+      const server = ChildProcess.spawn(config.command, args, config.options);
+      server.on('error', error => {
+        reject(error);
       });
+      server.on('close', () => () => {
+        this.server = null;
+      });
+      this.server = server;
+      this.port = port;
+
+      setTimeout(() => resolve(`http://localhost:${this.port}`), config.wait);
+    }));
   }
 
   stopServer() {
-    if (this.child) {
-      this.child.kill();
-      this.child = null;
+    if (this.server) {
+      this.server.kill();
+      this.server = null;
     }
+    return Promise.resolve();
   }
 
-  exitProcess() {
-    // generate a return value that is visible to the shell, 0 is success
-    this.process.exit(this.shellStatus);
+  /* eslint-disable no-process-exit */
+  exit(statusCode = 0) {
+    this.stopBrowser()
+      .then(() => this.stopServer())
+      .then(() => process.exit(statusCode))
+      .catch(error => {
+        this.logger.error(error.message || error);
+        process.exit(1);
+      });
   }
 
-  exit() {
-    return Promise.resolve()
-      .then(() => this.stopBrowser())
-      .then(() => {
-        this.stopServer();
-        this.exitProcess();
+  _getAvailablePort() {
```

Comment:
I'd put this util in a separate file. Not sure how Mac specific this, maybe a fallback?